Feat/worksheet nlq beta

noah-paige opened the pull request:
### Feature or Bugfix
<!-- please choose -->
- Feature
- Bugfix
- Refactoring

### Detail
- <feature1 or bug1>
- <feature2 or bug2>

### Relates
- <URL or Ticket>

### Security
Please answer the questions below briefly where applicable, or write `N/A`. Based on
[OWASP 10](https://owasp.org/Top10/en/).

- Does this PR introduce or modify any input fields or queries - this includes
fetching data from storage outside the application (e.g. a database, an S3 bucket)?
  - Is the input sanitized?
  - What precautions are you taking before deserializing the data you consume?
  - Is injection prevented by parametrizing queries?
  - Have you ensured no `eval` or similar functions are used?
- Does this PR introduce any functionality or component that requires authorization?
  - How have you ensured it respects the existing AuthN/AuthZ mechanisms?
  - Are you logging failed auth attempts?
- Are you using or adding any cryptographic features?
  - Do you use a standard proven implementations?
  - Are the used keys controlled by the customer? Where are they stored?
- Are you introducing any new policies/roles/users?
  - Have you used the least-privilege principle? How?


By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

noah-paige commented:
Pending Work:
- [ ] Add Integration Tests to Worksheets
- [ ] Add Documentation on how to use
- [ ] Testing!



Tests:

- [ ] Worksheets SQL Editor Query Owned Table
- [ ] Worksheets SQL Editor Query Shared Table
- [ ] Worksheets SQL Editor Query Shared Table w/ Filters

- [ ] Worksheets TextToSQL Editor Gen SQL Owned Table
- [ ] Worksheets TextToSQL Editor Gen SQL Shared Table
- [ ] Worksheets TextToSQL Editor Gen SQL Shared Table w/ Filters

- [ ] Worksheets DocSummarizer Editor List Objects Owned Dataset Bucket
- [ ] Worksheets DocSummarizer Editor Summarize Text Document

- [ ] Exceed Resource Invocation Threshold --> Throw Error
- [ ] Disable Feature Flags --> Disable API and UI Components

dlpzx commented on backend/dataall/base/db/exceptions.py:
```diff
@@ -181,3 +181,26 @@ def __init__(self, action, message):
+    def __str__(self):
+        return f'{self.message}'
+
+
```
ModelGuardrailException is used only once in `dataall.modules.worksheets.services.worksheet_service.WorksheetService.analyze_text_genai`, cannot we just raise a regular Exception?

noah-paige replied:
we can ... but thinking ahead if data.all was to support additional genAI features that were invoking models (ex: automated metadata generation) this custom exception could be re-used and be more helpful imo

dlpzx replied:
Convinced :)

dlpzx commented on backend/dataall/core/resource_threshold/db/resource_threshold.py:
```diff
@@ -0,0 +1,12 @@
+    actionId = Column(String(64), primary_key=True, default=utils.uuid('resource_threshold'))
+    username = Column(String(64), nullable=False)
+    actionType = Column(String(64), nullable=False)
+    date = Column(Date, default=date.today, nullable=False)
```
very nit: `date` is a field that is to understand at first sight. What date is it? We can add a comment or rename it so anyone understand what this column represents

noah-paige replied:
I'll change `actionId` to `actionUri` 

For `date` - what better name would you suggest?

dlpzx commented on backend/dataall/core/resource_threshold/db/resource_threshold_repositories.py:
```diff
@@ -0,0 +1,91 @@
+            .first()
+        )
+        return entry
+
```
check_invocation_count should be part of a service not of the db repository

dlpzx replied:
If we create the resoruce_threshold_service all the db functions will not need the prefix `_`

noah-paige replied:
I will move the decorator function to a service class

dlpzx commented on backend/dataall/core/resource_threshold/db/resource_threshold_repositories.py:
```diff
@@ -0,0 +1,91 @@
+                count = ResourceThresholdRepository._get_count_today(
+                    session=session, username=username, action_type=action_type
+                )
+                max_count = config.get_property(ResourceThresholdRepository._RESOURCE_PATHS[action_type], 10)
```
I am not sure we should be declaring the thresholds in the ResourceThresholdRepository._RESOURCE_PATHS. 
PROS: from the threshold configuration we can know all the possible configurable thresholds; CONS: to use `check_invocation_count` from any module we need to update the core module.

We can pass the parameter value to check_invocation_count(action_type, max_count_parameter_path) or directly the max_count value and call `config.get_property(ResourceThresholdRepository._RESOURCE_PATHS[action_type], 10)` directly in the worksheets module

noah-paige replied:
agreed with the above - I will add as an input parameter the config path

dlpzx commented on backend/dataall/modules/s3_datasets/aws/s3_dataset_client.py:
```diff
@@ -73,3 +73,21 @@ def get_bucket_encryption(self) -> (str, str, str):
+                    return True
+                else:
+                    return False
+
```
txt_or_pdf function is not needed:
`return [ob['Key'] for ob in response.get('Contents', []) if ob['Key'][ob['Key'].rfind('.')+1:] in ['pdf', 'txt']]`

Although for readability we might want to keep it as is

noah-paige replied:
it is essentially the same, right? I can make it a bit more clean

dlpzx commented on backend/dataall/modules/s3_datasets/api/dataset/resolvers.py:
```diff
@@ -156,6 +156,11 @@ def list_datasets_owned_by_env_group(
+def list_object_keys(context, source, datasetUri: str = None):
+    with context.engine.scoped_session() as session:
+        return DatasetService.list_object_keys(session=session, uri=datasetUri)
+
```
instead of using the context here an initializing the session in the resolver,  we should not pass the session and use get_context to start the session directly in the service after the permissions are checked

noah-paige replied:
yup oversight - fixing now

dlpzx commented on config.json:
```diff
@@ -77,7 +77,11 @@
+            "active": true,
+            "features": {
+                "nlq":true,
+                "max_count_per_day": 25
```
nit: move max_count_per_day inside the nlq feature? It makes it longer, but it is more organised
```
"nlq": {
    "active": true,
    "max_count_per_day": 25
},
```

noah-paige replied:
you are right I will change

dlpzx commented on backend/dataall/modules/worksheets/api/resolvers.py:
```diff
@@ -69,3 +70,49 @@ def run_sql_query(context: Context, source, environmentUri: str = None, workshee
+    databaseName: str = None,
+    tableNames: list = None,
+):
+    with context.engine.scoped_session() as session:
```
Same comment as before, let's avoid using the passed context object and start the session "too soon". Instead use get_context inside the service layer. Same for analyze_text_genai

noah-paige replied:
yup I will fix here and for other worksheet resolvers to match

dlpzx commented on backend/dataall/modules/worksheets/api/resolvers.py:
```diff
@@ -69,3 +70,49 @@ def run_sql_query(context: Context, source, environmentUri: str = None, workshee
+def text_to_sql(
+    context: Context,
+    source,
+    environmentUri: str = None,
```
why are these typing hints setting None as default? Aren't these fields non-nullable in gql?

noah-paige replied:
yup I will fix here and for other worksheet resolvers to match how the GQL Schema is defined

dlpzx commented on backend/dataall/modules/worksheets/services/worksheet_service.py:
```diff
@@ -1,5 +1,10 @@
 import logging
 
+from dataall.core.resource_threshold.db.resource_threshold_repositories import ResourceThresholdRepository
+from dataall.modules.worksheets.aws.glue_client import GlueClient
```
GlueClient: unused import

noah-paige replied:
only unused because I forgot to uncomment `run_nlq` method - should be used now 😄

dlpzx commented on backend/dataall/modules/worksheets/services/worksheet_service.py:
```diff
@@ -122,3 +129,55 @@ def run_sql_query(session, uri, worksheetUri, sqlQuery):
         )
 
         return AthenaClient.convert_query_output(cursor)
+
```
Why not adding `@TenantPolicyService.has_tenant_permission(MANAGE_WORKSHEETS)` to run_nlq?

noah-paige replied:
My understanding of `MANAGE_...` tenant level permissions is that they should be for mutating operations (i.e. creates, updates, deletes) on data.all but not for others

I think this has more to do with this open PR on re-working some of our Tenant Perms https://github.com/data-dot-all/dataall/pull/1630

let me know your thoughts - curious to think if you disagree

dlpzx replied:
Makes sense. Plus, the action is already protected by a resource permission which is more restrictive than a tenant permission

dlpzx commented on backend/dataall/modules/worksheets/services/worksheet_service.py:
```diff
@@ -122,3 +129,55 @@ def run_sql_query(session, uri, worksheetUri, sqlQuery):
+    @ResourcePolicyService.has_resource_permission(RUN_ATHENA_QUERY)
+    @ResourceThresholdRepository.check_invocation_count('nlq')
+    def run_nlq(session, username, uri, prompt, worksheetUri, db_name, table_names):
+        return 'TEST'
```
I think we need to un-comment this block of code

noah-paige replied:
yup - my final step was testing resource thresholds and didn't want to keep invoking the model 😄 - will uncomment

dlpzx commented on backend/dataall/modules/worksheets/services/worksheet_service.py:
```diff
@@ -122,3 +129,55 @@ def run_sql_query(session, uri, worksheetUri, sqlQuery):
+
+        content = s3_client.get_content(dataset.S3BucketName, key)
+        response = BedrockClient().invoke_model_process_text(prompt, content)
+
```
I think that the handling of the `invoke_model_process_text` error should happen inside the `invoke_model_process_text` itself

noah-paige replied:
this makes sense - good call out

dlpzx commented on backend/dataall/modules/worksheets/aws/bedrock_client.py:
```diff
@@ -0,0 +1,42 @@
+            'top_k': 250,
+            'top_p': 1,
+            'stop_sequences': ['\n\nHuman'],
+        }
```
possible typo: `self.__model` instead of `self._model`

dlpzx commented on backend/dataall/modules/worksheets/aws/bedrock_prompts.py:
I think having the prompts in a different file is a good start to organizing the prompts. But we can even make them easier to navigate by separating them in files inside a directory and use txt files tht do not require "", """"""
```
- bedrock_client.py
- bedrock_prompts/
     - text_to_sql_examples.txt
     - text_to_sql_template.txt
     - process_text_template.txt
```
And then use `PromptTemplate.from_file(template_file=PATH_TO_TXT_FILE)`

It was tricky to find an example, but int he unit_tests of langchain they use this [file](https://github.com/langchain-ai/langchain/blob/master/libs/core/tests/unit_tests/data/prompt_file.txt) in this [test](https://github.com/langchain-ai/langchain/blob/master/libs/core/tests/unit_tests/prompts/test_chat.py#L189)

This is just a suggestion, I am no genAI expert :)

noah-paige replied:
leaving until the end to pick up this change but I like the idea

dlpzx commented on backend/dataall/modules/worksheets/aws/bedrock_prompts.py:
```diff
@@ -0,0 +1,118 @@
+  ORDER BY price DESC
+  LIMIT 3)""",
+    },
+    {
```
nit: the second and third examples are pretty similar, maybe we can remove one of them to reduce the context sent to the LLM

noah-paige replied:
agreed - removing 3rd example

dlpzx commented on backend/dataall/modules/worksheets/aws/bedrock_prompts.py:
```diff
@@ -0,0 +1,118 @@
+    },
+    {
+        'User': """I want to see if any letter has been sent from 900 Somerville Avenue to 2 Finnigan Street what is the content n\nBased on on the following glue metadata: \n            
+    <context>
```
nit: are the typos on purpose? e.g. users that type fast? 
`I want to see if any letter has been sent from 900 Somerville Avenue to 2 Finnigan Street AND what is the content n\nBased ON ON the following glue metadata:`

noah-paige replied:
nope - an oversight ... fixing now

dlpzx commented on backend/dataall/modules/worksheets/aws/bedrock_prompts.py:
```diff
@@ -0,0 +1,118 @@
+
+Take the following points into consideration. It is crucial that you follow them:
+
+- I only want you to returned the SQL needed (NO EXPLANATION or anything else).
```
typo: to return

dlpzx commented on backend/dataall/modules/worksheets/aws/bedrock_prompts.py:
```diff
@@ -0,0 +1,118 @@
+- I only want you to returned the SQL needed (NO EXPLANATION or anything else).
+
+- Tables are referenced on the following form 'database_name.table_name' (for example 'Select * FROM database_name.table_name ...' and not 'SELECT * FROM table_name ...) since we dont have access to the table name directly since its not global vaiable.
+
```
typo: variable vaiable

dlpzx commented on backend/dataall/modules/worksheets/aws/bedrock_prompts.py:
```diff
@@ -0,0 +1,118 @@
+
+PROCESS_TEXT_PROMPT_TEMPLATE = """
+You are an AI assistant tasked with analyzing and processing text content. Your goal is to provide accurate and helpful responses based on the given content and user prompt. 
+You'll be given content to analyze and a prompt from the user based on the information in the document I want you to follow the following steps:
```
kind of typo: strange sentence flow: `You'll be given content to analyze and a prompt from the user based on the information in the document I want you to follow the following steps:`

noah-paige replied:
re-writing

dlpzx commented on backend/dataall/modules/worksheets/aws/bedrock_prompts.py:
```diff
@@ -0,0 +1,118 @@
+User prompt: {prompt}
+
+Please provide a response that addresses the user's prompt in the context of the given content. Be thorough, accurate, and helpful in your analysis.
+"""
```
In general I think we can be more concise with the prompts, remove redundancies and formulate the task in a clearer way

dlpzx commented on backend/migrations/versions/2258cd8d6e9f_resource_treshholds_added.py:
```diff
@@ -0,0 +1,37 @@
+"""describe_changes_shortly
```
nit: The name given in line 1 is the name that appears when we use alembic CLI commands, can we rename it to match the file name?

dlpzx commented on documentation/userguide/docs/worksheets.md:
```diff
@@ -55,4 +55,54 @@ delete buttons).
+As part of data.all >= v2.7 we introduced support for generative AI powered worksheet features. These features include both:
+
+1. Natural Language Querying (NLQ) of Structured Data
+2. Text Document Analysis of Unstructured Data
```
maybe good to mention that it is only available for txt as well as pdf files

noah-paige replied:
have a warning in the `Text Document Analysis of Unstructured Data` Section for this

dlpzx commented on documentation/userguide/docs/worksheets.md:
```diff
@@ -55,4 +55,54 @@ delete buttons).
+
+These features are optionally enabled/disabled via feature flags specified in data.all's configuration.
+
+The More details on how to use each of these features are below.
```
strange sentence: The More details

noah-paige replied:
removing "the"

dlpzx commented on documentation/userguide/docs/worksheets.md:
```diff
@@ -55,4 +55,54 @@ delete buttons).
+
+### Natural Language Querying (NLQ) of Structured Data
+
+data.all offers a NLQ feature to significantly redeuce the barrier for non-technical business users who need to quickly and easily query data to make informed decisions, as mastering these technical skills can be time-consuming.
```
typo: reduce

dlpzx replied:
not needed and a bit pessimistic: as mastering these technical skills can be time-consuming.

noah-paige replied:
removing and fixing

dlpzx commented on documentation/userguide/docs/worksheets.md:
```diff
@@ -55,4 +55,54 @@ delete buttons).
+
+data.all offers a NLQ feature to significantly redeuce the barrier for non-technical business users who need to quickly and easily query data to make informed decisions, as mastering these technical skills can be time-consuming.
+
+The NLQ feature will take a user prompt and select number of tables and generate a ready to run SQL statement that data.all user's can execute against the data they have access to in data.all's Worksheets module.
```
strange sentence: Given a prompt and a selection of tables, data.all NLQ will generate the corresponding SQL statement....

dlpzx commented on documentation/userguide/docs/worksheets.md:
```diff
@@ -55,4 +55,54 @@ delete buttons).
+
+The NLQ feature will take a user prompt and select number of tables and generate a ready to run SQL statement that data.all user's can execute against the data they have access to in data.all's Worksheets module.
+
+To start generating SQL, data.all user's can select the TextToSQL Tab in the Worksheets View:
```
typo in several places user's

dlpzx commented on documentation/userguide/docs/worksheets.md:
```diff
@@ -55,4 +55,54 @@ delete buttons).
+
+![worksheets_nlq](pictures/worksheets/ws_text_to_sql.png#zoom#shadow)
+
+
```
typo: From their

dlpzx commented on documentation/userguide/docs/worksheets.md:
```diff
@@ -55,4 +55,54 @@ delete buttons).
+![worksheets_nlq](pictures/worksheets/ws_text_to_sql.png#zoom#shadow)
+
+
+From their, user's will provide the environment, database, one or more tables to reference, and a user provided prompt. Behind the scenes' data.all will fetch the associated metadata for the table(s) the user selected, enrich the prompt with the additional metadata, and leverage genAI to generate a SQL statement that users can execute against their structured data.
```
typo scenes'

dlpzx commented on documentation/userguide/docs/worksheets.md:
```diff
@@ -55,4 +55,54 @@ delete buttons).
+### Text Document Analysis of Unstructured Data
+
+For unstructured text documents, data.all offers a feature to start analyzing your data using natural language.
+
```
typo: respwonse

dlpzx commented on documentation/userguide/docs/worksheets.md:
```diff
@@ -55,4 +55,54 @@ delete buttons).
+
+![worksheets_unstructured](pictures/worksheets/ws_analyze_txt_doc.png#zoom#shadow)
+
+
```
same typos as above: from their, user's

dlpzx commented on documentation/userguide/docs/worksheets.md:
The docs need a review to remove duplicated info, filler sentences and add examples

dlpzx commented on frontend/src/services/graphql/Datasets/listObjectKeys.js:
I would move this API call to the Worksheets module, it is only used there

dlpzx replied:
nit: the name is very open, maybe listS3ObjectKeys?

noah-paige replied:
currently the resolver logic is defined in `s3_dataset` module. Are you suggesting to move the entire FE and BE code to worksheets module?

Feels strange to do so when this API sole purpose is only to list keys --> more generic than worksheets

dlpzx replied:
There are 2 positions here:
- options 1: API in BE should be in the same spot in FE for coherence, this way developers can easily navigate both packages
- option 2: FE is just another client of BE, modularization of BE will not necessarily match the modularization of FE

I used to be more pro option-1; but the modularization design of the FE is following option 2: it groups services by the place they are used, so that the FE developer works in standalone FE modules with their views, components and services. 

But at the end of the day, the most important thing is that "we don't break the FE" which could happen if we disable a module in BE with API calls that affect a module in FE. For example, in this case, if we disable s3_datasets in BE, we are breaking worksheets in FE. The way to solve this problem for both options is that we define the dependency Worksheets-needs-S3Datasets. And this is implemented in both FE and BE.

dlpzx commented on frontend/src/modules/Worksheets/services/analyzeTextDocument.js:
```diff
@@ -0,0 +1,34 @@
+    environmentUri,
+    worksheetUri,
+    datasetUri: datasetUri,
+    key: key
```
why do we have a mix of kwargs and args here?

noah-paige replied:
making all positional - as we tend to do with other GQL Ops

dlpzx commented on tests_new/integration_tests/modules/s3_datasets/test_s3_dataset.py:
```diff
@@ -284,3 +285,14 @@ def test_persistent_import_kms_s3_dataset_update(client1, persistent_imported_km
+
+def test_list_object_keys(client1, persistent_s3_dataset1):
+    response = list_object_keys(client1, persistent_s3_dataset1.datasetUri)
+    assert_that(response).is_empty()
```
Is it always going to be empty?

noah-paige replied:
yes... until someone changes other tests to start uploading pdfs and txt files to the persistent_s3_dataset1 S3 Bucket 😄

dlpzx replied:
The time will come :)

dlpzx commented on tests_new/integration_tests/modules/worksheets/test_worksheet.py:
```diff
@@ -59,3 +63,83 @@ def test_update_worksheet_unauthorized(client2, worksheet1):
+        client2, prompt, persistent_env1.environmentUri, worksheet1.worksheetUri, ''
+    ).contains('UnauthorizedOperation', 'RUN_ATHENA_QUERY')
+
+
```
Commented code. I am not sure what we should do with this code block

noah-paige replied:
left it here to hear others opinions - not sure what is best approach is, will think on this and get back tomorrow with some thoughts

noah-paige commented on backend/dataall/modules/worksheets/aws/glue_client.py:
```diff
@@ -0,0 +1,36 @@
+        pivot_role_session = SessionHelper.remote_session(accountid=account_id, region=region)
+        aws_session = (
+            SessionHelper.get_session(base_session=pivot_role_session, role_arn=role) if role else pivot_role_session
+        )
```
yes - it is as such to be flexible for if there are other boto3 Glue APIs that need to be used at a later date that may not need to confine to this role chaining and just need pivot role

for the NLQ features this role will always be present because as you said we want to get the table metadata information as the team-IAM-role entity (or fail if we do not have access)

dlpzx commented on frontend/src/modules/Worksheets/components/TextDisplay.js:
```diff
@@ -0,0 +1,33 @@
+    maxWidth: '100%',
+    margin: '0 auto',
+    padding: '20px',
+    border:
```
Instead of using custom colors hardcoded, we could use the available palette (e.g. `backgroundColor: theme.palette.primary.main,` which already make the distinction between dark and light theme. The same applies to the font, instead of hardcoding Arial sans-serif use variable `fontFamily: theme.typography.fontFamily,`

dlpzx commented on frontend/src/modules/Worksheets/components/WorksheetDocAnalyzer.js:
```diff
@@ -0,0 +1,249 @@
+                    <MenuItem key={key} value={key}>
+                      {key}
+                    </MenuItem>
+                  ))
```
The list of S3 objects can be pretty long. As next steps we can add search functionality here. But next steps

noah-paige replied:
yes next step please 😄 - good enhancement though

dlpzx commented on frontend/src/modules/Worksheets/views/WorksheetView.js:
```diff
@@ -47,8 +43,32 @@ import {
+    value: 'SQLEditor',
+    active: true
+  },
+  {
```
It looks a bit ugly in the FE because instead of camelcase all letters are uppercase
![image](https://github.com/user-attachments/assets/ee41635d-3008-40c5-a694-47699d19edb9)

noah-paige replied:
Changing to `Text-To-SQL Editor`
![Screenshot 2024-10-24 at 9 26 14 AM](https://github.com/user-attachments/assets/f7504bf9-75ba-42ab-8eba-d2b3fa4a4c76)

dlpzx commented on frontend/src/modules/Worksheets/components/WorksheetTextToSQLEditor.js:
```diff
@@ -0,0 +1,241 @@
+              />
+            </Box>
+            <Box sx={{ p: 2 }}>
+              <LoadingButton
```
nit: disable button if it is going to throw an error. e.g. if environment, database and tables are not selected, disable it

noah-paige replied:
good call out - will do

dlpzx commented on frontend/src/modules/Worksheets/views/WorksheetView.js:
```diff
@@ -412,166 +467,55 @@ const WorksheetView = () => {
+              handleEnvironmentChange={handleEnvironmentChange}
+              loadingEnvs={loadingEnvs}
+              environmentOptions={environmentOptions}
+              currentEnv={currentEnv}
```
Nice, the selected environment and database is saved in all tabs :) :)

dlpzx reviewed (commented):
Reviewed and tested frontend and comments from yesterday

noah-paige commented:
Some example views from the UI:

Text To SQL:

![Screenshot 2024-10-24 at 11 55 50 AM](https://github.com/user-attachments/assets/7a4f3fe7-b2ed-4308-abbb-6da8233c68d7)


Document Analyzer:

![Screenshot 2024-10-24 at 11 55 33 AM](https://github.com/user-attachments/assets/96c2c524-d4b5-48dd-b9ba-5be9468dd643)

UserGuide Information in this PR:
`documentation/userguide/docs/worksheets.md` - changes in this PR


cc: @zsaltys @TejasRGitHub @anushka-singh

noah-paige commented:
@dlpzx - added additional tests and believe I have addressed all comments 

The one pending item is when testing the happy paths for both `TextToSQL` and `analyzeTextDocument` GQL operations - often the integration tests will fail with the following error:

`An error occurred (ThrottlingException) when calling the InvokeModel operation (reached max retries: 4): Too many requests, please wait before trying again. You have sent too many requests.  Wait before trying again.`

If you retry the test in isolation / after some time it will resolve itself. Looking at the service quota limits, Claude 3.5 Sonnet has a fairly low max invoke count per account per minute of 50 (note: we are still far under that per minute threshold but I think that or similar could be causing this throttling exception - other Claude 3 models have a much higher per minute tolerance of invocations)

Not sure if we should opt for a Claude 3 version for the time being to remediate this testing case and also to improve UX for a user using these features, thoughts?

dlpzx commented on backend/dataall/modules/worksheets/aws/bedrock_prompts/process_text_template.txt:
```diff
@@ -0,0 +1,14 @@
+You are an AI assistant tasked with analyzing and processing text content. Your goal is to provide accurate and helpful responses based on the given content and user prompt. 
+You must follow the steps:
+
+1. Detetermine if the document has the information to be able to answer the question. If not respond with "Error: The Document does not provide the information needed to answer you question"
```
typo: you question -> your question

dlpzx commented on backend/dataall/modules/worksheets/services/worksheet_service.py:
```diff
@@ -122,3 +150,48 @@ def run_sql_query(session, uri, worksheetUri, sqlQuery):
+        glue_client = GlueClient(
+            account_id=environment.AwsAccountId, region=environment.region, role=env_group.environmentIAMRoleArn
+        )
+
```
ultra nit: the only metadata that we send are the table names no? I would name the variable table_names instead of metadata so that we know what we are sending as LLM context

noah-paige replied:
We send the following for each table selected to the LLM: 
```
            Database Name: {database}
            Table Name: {table_name} 
            Column Metadata: {column_metadata}
            Partition Metadata: {partition_metadata}
```

dlpzx reviewed (approved):
Looks good, but needs to resolve conflicts

dlpzx commented on backend/dataall/modules/worksheets/aws/bedrock_client.py:
```diff
@@ -0,0 +1,51 @@
+    def __init__(self):
+        self._session = SessionHelper.get_session()
+        self._client = self._session.client('bedrock-runtime')
+        model_id = 'anthropic.claude-3-5-sonnet-20240620-v1:0'
```
I am facing errors of the type: ` An error occurred (ValidationException) when calling the InvokeModel operation: Invocation of model ID anthropic.claude-3-5-sonnet-20240620-v1:0 with on-demand throughput isn’t supported. Retry your request with the ID or ARN of an inference profile that contains this model.` I think it is related to the enforced [cross-region inference](https://docs.aws.amazon.com/bedrock/latest/userguide/cross-region-inference.html)

dlpzx replied:
I wonder if you have faced something similar when using this model
